### ISSUE TITLE
Feature request: Pause / Resume for Lawnmower Mode

### ISSUE THREAD

**Issue by EmKarr** [biigle/core#547]
Dear Martin,

whilst scanning a large image with the lawnmower mode to make sure we do not miss anything, we often need to zoom in for investigating details of individual objects. Alas, lawnmower mode and zooming do not go well together and cause the lawnmower to reposition the current field of view erratically (or if not erratically, at least not in a helful way).

Our suggestion to solve this issue (and bring some other benefit) would be to have a pause / resume function for the lawnmower mode: Interrupt it either with a specific keyboard shortcut or (optionally, but nice to have) automatically when the view is zoomed in or out, and then continue it at the exact position and zoom level used before pausing with another specific keyboard shortcut. 

Another nice featere would be to be able to optionally start the lawnmower mode from the current field of view instead of the lower left corner, to be able to continue analysing after a long break or in another session, or even to log the progress and resume from the last used position.

Thank you very much for BIIGLE any our constant support,
Michael

**Comment by mzur** [comment]
Hi Michael, thanks for the ideas. That's certainly something that could be added.

Here come my technical notes for when someone picks this up for implementation:
Change the "power" icon on the button to start Lawnmower Mode to a "play" icon. The function stays the same. However, when Lawnmower Mode is active, the icon switches to a "stop" icon. Next to the button, add a new "pause" button that can be clicked when Lawnmower Mode is active. When this button is clicked, the current image, viewport and zoom is remembered in local storage. The "pause" button gets the active state and the "stop" icon will revert to the "play" icon (see below). When the "play" button is pressed again, Lawnmower Mode will continue at the stored image, viewport and zoom (even if the user currently views another image of the volume). If the "pause" button is clicked while it is active, the stored state will be forgotten. Each "pause state" will be remembered separately for each volume.

Also, change all the other "power" buttons of the annotation modes to "play"/"stop" to be consistent.

This is how it could look like:

![Screenshot from 2023-02-17 09-37-45](https://user-images.githubusercontent.com/2457311/219594898-4e3cf722-10cb-4ba6-8ad2-4ddfef745d8d.png)
(Lawnmower Mode is off, no pause state available)

![Screenshot from 2023-02-17 09-36-34](https://user-images.githubusercontent.com/2457311/219595200-247a0593-278d-4533-a8bc-faeabd4f2899.png)
(Lawnmower Mode is on, can be paused)

![Screenshot from 2023-02-17 09-36-56](https://user-images.githubusercontent.com/2457311/219594971-3eff9d1e-78b9-4752-8b96-c6287e5ef9da.png)
(Lawnmower Mode is paused, can either be resumed or pause state can be forgotten)

**Comment by yannik131** [comment]
@mzur I'm testing the lawnmower mode right now, it's not supposed to jump to the center of the next/previous image like this, right? It's just skipping the first/last tile this way. I'll first fix this, then add the pause/play functionality. I assume the current behaviour of jumping to the nearest tile when changing the zoom while lawnmower mode is active should be kept?

https://github.com/user-attachments/assets/2cc9976c-958c-46a6-98e5-97c1846baa15

**Comment by mzur** [comment]
> I'm testing the lawnmower mode right now, it's not supposed to jump to the center of the next/previous image like this, right?

On biigle.de I don't see this behavior. But do have a look.

>  I assume the current behaviour of jumping to the nearest tile when changing the zoom while lawnmower mode is active should be kept?

Yes.

**Comment by yannik131** [comment]
@mzur 

> Each "pause state" will be remembered separately for each volume.

Can you elaborate on that? Do you want to store the pause state for each volume in local storage, so that if the page is reloaded or navigated away from and then back to it, the user can still resume?

Edit: I don't think this even makes sense, see following post. Navigating away from page always stops lawnmower, so tying the information to the volume is unnecessary I think.

**Comment by yannik131** [comment]
@mzur to summarize, we want this behavior for the various state transitions, right?

- states: active, inactive, paused
- initial state: inactive
- inactive -> active: activate lawnmower as before
- active -> inactive: deactivate like before
- active -> paused: deactivate + save state
- paused -> active: activate + load saved state
- paused -> inactive: discard state

We still need to discuss how any of these interruptive actions should be handled when lawnmower mode is active:
1. reloading the page, navigating away from and back to the page: don't do anything, this stops lawnmower, no save
2. switching to volare or regular/random sampling mode: stops lawnmower, no save
3. change viewport (center or zoom). OP says the original behavior of jumping to the nearest tile bothered him, but you said we should keep the original behavior. I'd propose to treat this as active -> paused instead, saving the current state.

Also, do we want a keybind for pause/resume? It could be the same for volare and lawnmower, depending on what is active.

**Comment by mzur** [comment]
I think my original idea with storing the info in local storage was that the user can start annotating lots of images and pause at some point. Then they come back the next day and can resume where they left off.

But I see now that this is not the same use case than described by OP. They just want a way to zoom in without disrupting the annotation mode. This is already possible with volume label review mode, so the feature without local storage would not add much value. Maybe we can serve both use cases.

What about this: The annotation mode is only paused if the user explicitly hits the pause button or they change the viewport. Reloading the page or switching to another annotation model does not enable the paused mode.

Maybe we can also store a timestamp so if the mode is paused we can add "paused on xxx at xxx". What do you think?

**Comment by yannik131** [comment]
> This is already possible with volume label review mode, so the feature without local storage would not add much value. Maybe we can serve both use cases.

True, but trying to navigate to another image with an arrow button immediately activates volare again. A pause state would prevent this. 

> What about this: The annotation mode is only paused if the user explicitly hits the pause button or they change the viewport. Reloading the page or switching to another annotation model does not enable the paused mode.

What about leaving the page when in paused mode, should the state be reloaded from local storage when the user returns to the volume later? 

> Maybe we can also store a timestamp so if the mode is paused we can add "paused on xxx at xxx". What do you think?

Do you mean showing this in the canvas, as a toast or next to the mode toggle in the sidebar? And then "paused on image X at 15:00:00"? I'm not sure if users care about that.. You probably have more experience annotating things than I do 😄

**Comment by mzur** [comment]
> What about leaving the page when in paused mode, should the state be reloaded from local storage when the user returns to the volume later?

No.

> Do you mean showing this in the canvas, as a toast or next to the mode toggle in the sidebar? And then "paused on image X at 15:00:00"? I'm not sure if users care about that.. You probably have more experience annotating things than I do 😄

I meant to just show it as (muted) text next to the buttons in the tab. Not on the canvas and not as a toast. It may be useful if users discover the paused state and have to decide if they can dismiss it (e.g. it was activated months ago).

**Comment by yannik131** [comment]
@mzur Okay so as far as I understand you want:
- the pause state of volare and lawnmower is saved volume specific in local storage
- pause state is entered by either clicking pause or changing the viewport while volare or lawnmower is active
- when the page reloads, the pause state, if present, is reloaded from local storage (otherwise your "activated months ago" statement wouldn't be possible)
- a muted text next to the play/pause buttons indicates when it was paused (I suggest "paused X Y ago" where Y is seconds, minutes, hours, days, ... and X is the amount, otherwise we'd need a full datetime stamp to accomodate "activated months ago" which might be too large)

Is that correct?

**Comment by mzur** [comment]
Sounds right 👍
